### PR TITLE
New commands for Orca Integration

### PR DESCRIPTION
**Pull request by viktor-orca**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by edik24** [comment]
@viktor-orca please fill the contribution form here:
https://github.com/demisto/content/pull/25561#issuecomment-1485580648

**Comment by merit-maita** [comment]
Hi @Victor-Yurievich ,
Thank you for your contribution.
Due to holidays in Israel in the upcoming days, your PR review will be slightly delayed, I apologize in advance.
In the meantime, please fill in the following [registration form](https://forms.gle/XDfxU4E61ZwEESSMA)
You can also try to solve the errors in [this build](https://circleci.com/gh/demisto/content/694608)
I’m available for any of your questions here or on slack.
Thanks again.

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 1]
```diff
@@ -237,6 +307,101 @@ def fetch_incidents(
...
+            "severity": response.get("details", {}).get("severity")
+        }
+    if "error" in response:
+        context = {
+            "details": response.get("error")
+        }
+
+    return CommandResults(
+        readable_output="Alert severity changed",
+        outputs_prefix="Orca.SetAlertSeverity",
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.py:333]: when updating an entity (like the alert in your case), no need to set an output key for that (SetAlertSeverity), since its not an entity itself, its just a change, so a human readable, a message it was changed and what is the change, should be enough, for the output field, you can use the Orca.Alert key (since its a key) .

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 2]
```diff
@@ -237,6 +307,101 @@ def fetch_incidents(
...
+def set_alert_status(orca_client: OrcaClient, args: Dict[str, Any]) -> CommandResults:
+    alert_id = cast(str, args.get("alert_id"))
+    status = cast(str, args.get("status"))
+    assert alert_id and status
+    demisto.debug(f"Set alert status {alert_id=} {status=}")
+
+    orca_client.set_alert_status(alert_id=alert_id, status=status)
+    return CommandResults(
+        readable_output="Set alert status",
+        outputs_prefix="Orca.SetAlertStatus",
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.py:373]: same comment for Orca.SetAlertSeverity, please address here as well

File: Packs/Orca/Integrations/Orca/Orca.yml [hunk 3]
```diff
@@ -1,117 +1,210 @@
...
+      description: Change severity for the alert
+      arguments:
+        - name: alert_id
+          description: Alert id
+          required: true
+        - name: score
+          description: The Orca Score to set. Valid range 1.0 to 10.
+          required: true
+      outputs:
+        - contextPath: Orca.SetAlertSeverity.id
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.yml:124]: same comment i mentioned in the code:
when updating an entity (like the alert in your case), no need to set an output key for that (SetAlertSeverity), since its not an entity itself, its just a change, so a human readable, a message it was changed and what is the change, should be enough, for the output field, you can use the Orca.Alert key (since its a key) .

File: Packs/Orca/Integrations/Orca/Orca.yml [hunk 4]
```diff
@@ -1,117 +1,210 @@
...
+          predefined:
+            - in_progress
+            - dismiss
+            - dismissed
+            - open
+            - close
+            - closed
+          description: Alert status
+      outputs:
+        - contextPath: Orca.SetAlertStatus
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.yml:186]: same comment for Orca.SetAlertSeverity

**Review by merit-maita — CHANGES_REQUESTED** [review]
Thank you so much for your contribution, it looks great!
please address my comments, for any questions i'm available over here or slack

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 5]
```diff
@@ -237,6 +307,101 @@ def fetch_incidents(
...
+    response = orca_client.set_alert_score(alert_id=alert_id, orca_score=score)
+
+    context = {}
+    if "alert_id" in response:
+        context = {
+            "id": response.get("alert_id"),
+            "details": response.get("details", {}).get("description"),
+            "severity": response.get("details", {}).get("severity")
+        }
+    if "error" in response:
```
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:326]: We handle error here because api will return error if the alert have the same severity. 
Not sure that we should raise exception in this case.

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 6]
```diff
@@ -27,9 +30,16 @@ def validate_api_key(self) -> str:
                 url_suffix="/rules/query/alerts", data={},
                 timeout=ORCA_API_TIMEOUT
             )
-        except Exception:
-            return invalid_token_string
-        if response.get("status") != "success":
+            if response.get("status") != "success":
+                return response.get("error") or invalid_token_string
```
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:34]: This logic is needed to improve the validation error message.
The API returns the reason why the validation failed, and it would be good to show this message to the user.

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 7]
```diff
@@ -52,8 +62,8 @@ def get_alerts_by_filter(
         params = {}
         if alert_type:
             params = {"type": alert_type}
-        elif asset_unique_id:
-            params = {"asset_unique_id": asset_unique_id}
+        else:
+            params = {"asset_unique_id": asset_unique_id}  # type: ignore[dict-item]
```
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:66]: It's not important for the API. It will works in both cases. 
I just revert the changes

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:34]: This custom error handler is needed to display the correct error message to the user
Example: image

Not sure how to do it with the return_error() function.

[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca.py:34]: I'm sorry, i don't understand your point

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca.py:327]: what is the message returned in the error from the response? i see why it's not considered as a failure, but we we're expecting to see?

File: Packs/Orca/Integrations/Orca/Orca_test.py [hunk 8]
```diff
@@ -635,3 +631,141 @@ def test_fetch_all_alerts(requests_mock, orca_client: OrcaClient) -> None:
...
+            "description": "Alert risk level changed",
+            "severity": "Hazardous"
+        }
+    })
+
+    response = set_alert_severity(orca_client=orca_client, args={
+        "alert_id": alert_id,
+        "score": 6
+    })
+    assert response.to_context()['Contents']['alert_id'] == alert_id
```
[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca_test.py:657]: what about this?

File: Packs/Orca/Integrations/Orca/Orca_test.py [hunk 9]
```diff
@@ -635,3 +631,141 @@ def test_fetch_all_alerts(requests_mock, orca_client: OrcaClient) -> None:
...
+            }
+        }
+    )
+    args = {
+        "alert_id": "orca-1",
+        "status": "open"
+    }
+    result = set_alert_status(orca_client, args)
+    content = result.to_context()["Contents"]
+    assert content["status"]
```
[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca_test.py:728]: and this?

**Review by merit-maita — COMMENTED** [review]
it looks very good overall, just please address my comments for the unit tests and 2 other added comments.
thank you!!

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:327]: In case if the alert already have requested severity - response will have status code 400 and `error` will contains `requested to set same configuration`
By my opinion it is not error from the user perspective.

[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:328]: Anyway i changed custom handler. Instead i just raise an DemistoException

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py]: changed to use `return_error()`

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca.py:328]: no, i agree that in this case it should not be considered as an error from the user perspective, so please add an if statement to check if this error is returned (the same state) then it should not throw an error just return a statement that the "the stats is ...."
otherwise, which's a real error, it should throw an exception.

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (merit-maita) [reply Packs/Orca/Integrations/Orca/Orca.py]: did it work as you expected it to work after change?

**Review by merit-maita — COMMENTED** [review]
just one minor comment added,
and your unit tests checks are failing please make sure to handle it
thanks!!

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py]: Yes, I've checked it `return_error(message="")` works is i expected.

**Review by merit-maita — APPROVED** [review]

**Comment by merit-maita** [comment]
@viktor-orca looks very good, thank you for the effort!
when does it suit you to set up a demo?

**Comment by viktor-orca** [comment]
> @viktor-orca looks very good, thank you for the effort! when does it suit you to set up a demo?

@merit-maita Thank you for you review. 
What is the format of the demo?
Should i prepare Playbook or just run commands in Playground?

**Comment by merit-maita** [comment]
@viktor-orca its really up to you, the goal is to see that everything is working as expected with your changes, along with your changes, please let me know when it works for you

**Comment by viktor-orca** [comment]
> @viktor-orca its really up to you, the goal is to see that everything is working as expected with your changes, along with your changes, please let me know when it works for you

@merit-maita Before the demonstration, I would like to ask for a new license for our test server. Otherwise it's not possible to demo new commands without license :)

**Comment by merit-maita** [comment]
@viktor-orca 
@edik24 is taking care of it.
for the demo, i will be away all next week for the holidays, i'll be back on 30/4 Sunday,
sorry for the delay!

**Comment by viktor-orca** [comment]
> taking care of it.
> for the demo, i will be away all next week for the holidays, i'll be back on 30/4 Sunday,
> sorry for the delay!

Thanks @merit-maita 
If you have time I can make demo today.

**Comment by merit-maita** [comment]
as discussed in slack, we can't today, so let's set it for next week after 30/4

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 10]
```diff
@@ -140,24 +151,91 @@ def get_asset(self, asset_unique_id: str) -> Union[Dict[str, Any], str]:  # pyli
...
     demisto_informational = 0.5
-    # demisto_low = 1  (commented because of linter issues)
+    demisto_low = 1
     demisto_medium = 2
     demisto_high = 3
     demisto_critical = 4
 
     # LHS is Orca score
-    MAPPING = {1: demisto_critical, 2: demisto_high, 3: demisto_medium, 4: demisto_informational}
+    MAPPING = {
```
Comment (viktor-orca) [inline Packs/Orca/Integrations/Orca/Orca.py:230]: @merit-maita Please check this changes. 
It was changed after your review.

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 11]
```diff
@@ -169,15 +247,15 @@ def get_incident_from_alert(alert: Dict[str, Any]) -> Dict[str, Any]:
         'name': alert.get('state', {}).get('alert_id'),
         'occurred': last_seen_time,
         'rawJSON': json.dumps(alert),
-        'severity': map_orca_score_to_demisto_score(orca_score=alert.get('state', {}).get('score'))
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.py:172]: why did you change it from score to risk_level?

**Comment by merit-maita** [comment]
please review the failing validation check

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 11 above)
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:172]: `score` will deprecated soon.

**Comment by merit-maita** [comment]
Post Demo:
@viktor-orca thank you again for your contribution 
here's regarding the points we mentioned in demo:
- for how to display the data for the event log command: see this [example](https://github.com/demisto/content/blob/ae19021c83ccedec87ad407345ed0f79dc9aff9f/Packs/HostIo/Integrations/HostIo/HostIo.py#L119)
- to change the readable_output section for commands to be more clear.

File: Packs/Orca/ReleaseNotes/2_2_0.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+<!--
+#### Integrations
+##### Orca Event Collector
+  - No changes
+
+#### Modeling Rules
+##### Orca Modeling Rules
+  - No changes
+-->
```
[the remark this answers is not in the document]
Comment (viktor-orca) [reply Packs/Orca/ReleaseNotes/2_2_0.md:21]: I added it for the fix validation check.
Here related xsoar documentation https://xsoar.pan.dev/docs/documentation/release-notes#excluding-items

**Comment by viktor-orca** [comment]
> Post Demo: @viktor-orca thank you again for your contribution here's regarding the points we mentioned in demo:
> 
> * for how to display the data for the event log command: see this [example](https://github.com/demisto/content/blob/ae19021c83ccedec87ad407345ed0f79dc9aff9f/Packs/HostIo/Integrations/HostIo/HostIo.py#L119)
> * to change the readable_output section for commands to be more clear.

@merit-maita Thanks for your comments and review.

The output fixed.

File: Packs/Orca/Integrations/Orca/Orca.py [hunk 13]
```diff
@@ -237,6 +315,108 @@ def fetch_incidents(
...
+
+    alert_logs = [{
+        "id": item.get("id"),
+        "alert_id": item.get("alert_id"),
+        "type": item.get("type"),
+        "description": item.get("details", {}).get("description")
+    } for item in context]
+
+    return CommandResults(
+        readable_output=tableToMarkdown(f"Alert event log ({alert_id})", alert_logs, removeNull=True),
```
Comment (merit-maita) [inline Packs/Orca/Integrations/Orca/Orca.py:371]: did you check if we're getting results when running the command?

**Review by merit-maita — COMMENTED** [review]
it looks. great, thank you!
did we get the result we're expecting for event logs?

File: Packs/Orca/Integrations/Orca/Orca.py (hunk 13 above)
Comment (viktor-orca) [reply Packs/Orca/Integrations/Orca/Orca.py:371]: Yes, sure
image
